src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala

```diff
@@ -8,17 +8,103 @@
 package no.gdl.bookapi.service.translation
 
 import com.typesafe.scalalogging.LazyLogging
-import no.gdl.bookapi.model.api.{TranslateRequest, TranslateResponse}
+import io.digitallibrary.language.model.LanguageTag
+import no.gdl.bookapi.integration.crowdin.{CrowdinClient, CrowdinClientBuilder, CrowdinUtils}
+import no.gdl.bookapi.model._
+import no.gdl.bookapi.model.api._
+import no.gdl.bookapi.model.domain.{InTranslation, InTranslationFile, TranslationStatus}
+import no.gdl.bookapi.service.ReadService
 
-import scala.util.{Success, Try}
+import scala.util.{Failure, Success, Try}
 
 trait TranslationService {
+  this: CrowdinClientBuilder with ReadService with SupportedLanguageService with TranslationDbService =>
   val translationService: TranslationService
 
+
   class TranslationService extends LazyLogging {
-    def addTranslation(translateRequest: TranslateRequest): Try[TranslateResponse] = {
-      // TODO in #155: Add implementation of sending to crowdin
-      Success(TranslateResponse(1, "https://www.digitallibrary.io"))
+    def fetchTranslationsIfAllTranslated(projectIdentifier: String, language: LanguageTag): Try[Unit] = {
+      Success()
+      // TODO in #138: Implement fetch translations
+    }
+
+    def updateTranslationStatus(projectIdentifier: String, language: LanguageTag, fileId: String, status: TranslationStatus.Value): Try[InTranslationFile] = {
+      translationDbService.fileForCrowdinProjectWithFileIdAndLanguage(projectIdentifier, fileId, language) match {
+        case None => Failure(new NotFoundException(s"No translation for project $projectIdentifier, language $language and file_id $fileId"))
+        case Some(file) =>
+          translationDbService.updateTranslationStatus(file, status)
+      }
+    }
+
+    def addTranslation(translateRequest: api.TranslateRequest): Try[api.TranslateResponse] = {
+      Try(LanguageTag(translateRequest.fromLanguage)).flatMap(fromLanguage => {
+        validateToLanguage(translateRequest.toLanguage).flatMap(toLanguage => {
+          readService.withIdAndLanguage(translateRequest.bookId, fromLanguage) match {
+            case None => Failure(new NotFoundException())
+            case Some(originalBook) => {
+              crowdinClientBuilder.forSourceLanguage(fromLanguage).flatMap(crowdinClient => {
+                val existingTranslations = translationDbService.translationsForOriginalId(translateRequest.bookId)
+                val toAddUser = existingTranslations.find(tr => tr.fromLanguage == fromLanguage && tr.toLanguage == LanguageTag(toLanguage))
+                val toAddLanguage = existingTranslations.find(tr => tr.fromLanguage == fromLanguage)
```

Comment:
Are these two mutually exclusive? It looks like this might be too coupled with the `if` logic below.

Reply:
Yes, they are.
Have replaced the if-tests with a match like:
```
val inTranslation = (toAddUser, toAddLanguage) match {
   case (Some(addUser), _) => addUserToTranslation(addUser)
   case (None, Some(addLanguage)) => addTargetLanguageForTranslation(addLanguage, translateRequest, crowdinClient)
   case _ => createTranslation(translateRequest, originalBook, fromLanguage, toLanguage, crowdinClient)
}
```

So the ordering of the previous if-test is goen 😄